Modified Submission Makefile for Uni Seminar

timkicker opened the pull request:
I created an improved version of the submission Makefile used for our uni seminar and thought i may share it. The key improvements are as follows:

- **Code formatting:**  
  Formats all *.c files via the provided clang file.

- **Automatic `group.txt` Creation:**  
  The Makefile now automatically creates a `group.txt` file in the exercise folder, depending on the `PEOPLE` variable (see below). This file is then included in the submission zip.

- **Submission Folder Creation:**  
The Makefile checks if a `submission` folder exists and creates it if necessary, ensuring that the zip files are placed in a unified location.

- **Conditional Archive Creation:**  
Depending on the `PEOPLE` variable, the Makefile creates either a single archive (if `PEOPLE=1`) or two archives (if `PEOPLE=2`), which allows for both individual and group submissions.

- **World-Readable Permissions:**  
A new target sets all files in the exercise folder to be world-readable (`chmod a+r`), ensuring that third parties can read the files if needed.

## How to Use
- Run `make all` to execute the complete workflow (clean, group file creation, formatting, permission setting, and zip archive creation).
- Use the `PEOPLE` variable to control the number of archives:
- `make PEOPLE=1 all` creates only the archive for Person 1.
- `make PEOPLE=2 all` (or simply `make all`, as the default is 2) creates both archives.

timkicker committed:
modify general makefile

GagaLP commented on Makefile:
```diff
@@ -0,0 +1,93 @@
+.PHONY: format
+format:
+	@echo "Formatting all .c files in exercise$(EXERCISE)..."
+	@find exercise$(EXERCISE) -type f -name "*.c" -exec clang-format -i {} \;
```
This is a great addition. However, we are concerned that it may not work for students who do not have the LLVM toolchain installed.

There are two possible ways to proceed:
1. Remove it, which would be unfortunate since it could help with formatting.
2. Before execution, check if `clang-format` is installed on the system.

GagaLP commented on Makefile:
```diff
@@ -0,0 +1,93 @@
+
+.PHONY: group
+group:
+ifeq ($(strip $(PEOPLE)),1)
```
Group creation is definitely a useful feature for students, but it is too verbose and makes the Makefile less readable.

To improve this, use a `for` loop to enhance readability and make the Makefile cleaner.

GagaLP commented on Makefile:
```diff
@@ -0,0 +1,93 @@
+zip: prepare
+	@mkdir -p submission
+	$(RM) ./submission/$(ARCHIVE_PERSON_1)
+ifeq ($(strip $(PEOPLE)),1)
```
Similar to the above, use a `for` loop here to improve readability and keep the code cleaner.

GagaLP reviewed (changes requested):
Thanks for this pull request! We appreciate and support student engagement. While these changes add useful content, there are a few things to keep in mind.  

I have reviewed your changes and added some comments that need to be addressed before we can merge this.  

### General pointers for this kind of PR:  
- Keep in mind that not all students have the same setup as you. Some may lack the necessary tools to execute this Makefile. Therefore, keep dependencies as low as possible. If you need to add a dependency like `clang-format`, always check if it is installed before running the command, and if it's missing, handle it gracefully rather than terminating the script abruptly.  
- You did an excellent job describing the Makefile in the PR, but there is little documentation within the Makefile itself. Add some general explanations of what the variables and functions do.  
- The main goal of this Makefile is to help students understand how Makefiles work. If the script becomes too verbose, it will be harder to follow. Please keep verbosity to a minimum.

timkicker commented:
Thanks for the quick response!  
I also agree that encouraging student engagement is really important.  
Optimizing the Makefile for learning purposes sounds like a good idea — I'll work on the improvements over the next few days (or during the week).  
Thanks again!

timkicker committed:
add clang check for makefile and refactor

timkicker commented:
So, I went through the suggestions and updated the Makefile accordingly:
- The individual steps are now commented and cleaned up to be easy to understand for students.
- The archive creation for zips uses a loop based on `PEOPLE`, no more repetitive code.
- I tried to use the same procedure on the group creation but could'nt get it to work in a way that I'm satisfied, so i tried it in a way that's at least readable and understandable by others
- `clang-format` is now checked before use, and skipped with a message if not available.
- Improved overall readability by minimizing verbosity and adding clear comments explaining each major part.

I'd be happy to get your feedback on this version too, especially if you think there are more ways to make it even cleaner

GagaLP commented on Makefile:
```diff
@@ -0,0 +1,113 @@
+
+# Number of people: set to 1, 2 or 3 (how many zips to create)
+PEOPLE ?= 3
+USERNAME_PERSON_1 = csbaXXXX
```
Make those an array

GagaLP commented on Makefile:
```diff
@@ -0,0 +1,113 @@
+EXERCISE = 05
+
+# Data for group.txt. Only relevant if PEOPLE is greater than 1.
+PERSON_1 = Max Mustermann
```
Similar with those

GagaLP commented on Makefile:
```diff
@@ -0,0 +1,113 @@
+PERSON_1 = Max Mustermann
+PERSON_2 = Gordon Freeman
+PERSON_3 = Alyx Vance
+MAT_NUM_1 = 12345678
```
Same

GagaLP commented on Makefile:
```diff
@@ -0,0 +1,113 @@
+else
+	@echo "Creating group.txt in exercise$(EXERCISE)..."
+	@rm -f exercise$(EXERCISE)/group.txt
+	@for i in 1 2 3; do \
```
When you include an array, as I mentioned above, you can loop over the array instead of this here, removing the need of the `for` `if` construct.

GagaLP commented on Makefile:
```diff
@@ -0,0 +1,113 @@
+	@for i in 1 2 3; do \
+	    if [ $$i -le $(PEOPLE) ]; then \
+	        case $$i in \
+	            1) echo "$(MAT_NUM_1) $(PERSON_1)" >> exercise$(EXERCISE)/group.txt ;; \
```
This approach is generally not ideal, as it doesn't fully leverage the purpose of `for` loops.  
`For` loops are meant to help avoid code duplication, but using a `switch` statement here results in similar repetition as the previous `if-else` structure.  
Consider using arrays to make this part cleaner and more efficient.

GagaLP reviewed (changes requested):
Sorry for the late review, and thanks again for your work.

Some parts are still not quite at the level we'd consider ready to merge, but most of the changes are definitely heading in the right direction.

As a general pointer: try to avoid code repetition and the use of large numbers of individual variables. Consider using arrays, as they can help reduce duplication and make the code cleaner.
When adding new functionality, we always aim to make changes as general as possible — for example, instead of limiting the number of people to 3, make it scalable to N people. This will make the code more flexible and future-proof in case we want to support larger groups later on.

timkicker committed:
Refactor submission Makefile: use arrays and loops for scalability

timkicker commented:
Thanks again for the feedback. Also no worries at all about the timing, noone expects you to do free work, especially during easter-break ;) . I’ve just pushed a refactor where:

- All user data (usernames, names, matriculation numbers) live in arrays  
- The `group.txt` and ZIP creation now loop dynamically over `PEOPLE`  
- Names are stored with underscores and then converted to spaces when writing `group.txt`  
- Also did some refactoring

Could you have another look to see if this meets the goals?

stefanpedratscher commented on Makefile:
```diff
@@ -0,0 +1,97 @@
+
+# Names and matriculation numbers for group.txt (only used if PEOPLE > 1)
+# Underscores in PERSONS are replaced by spaces in the output.
+PERSONS   = Max_Mustermann Gordon_Freeman Alyx_Vance
```
```suggestion
PERSONS = Max_Mustermann Gordon_Freeman Alyx_Vance
```

stefanpedratscher commented on Makefile:
```diff
@@ -0,0 +1,97 @@
+# Names and matriculation numbers for group.txt (only used if PEOPLE > 1)
+# Underscores in PERSONS are replaced by spaces in the output.
+PERSONS   = Max_Mustermann Gordon_Freeman Alyx_Vance
+MAT_NUMS  = 12345678 87654321 98765432
```
```suggestion
MAT_NUMS = 12345678 87654321 98765432
```

stefanpedratscher commented on Makefile:
```diff
@@ -0,0 +1,97 @@
+MAT_NUMS  = 12345678 87654321 98765432
+
+# Exercise number / folder
+EXERCISE  = 04
```
```suggestion
EXERCISE = 04
```

stefanpedratscher reviewed (commented):
Hi @timkicker and @GagaLP ,

thank you both very much for the changes and the effort you put into improving the Makefile. I’ve reviewed it and believe it’s ready to be merged. I’ve suggested a few minor adjustments to improve consistency, but overall everything looks good.

Thanks again!

timkicker committed:
Update Makefile

Co-authored-by: Stefan Pedratscher <pedratscher.stefan@gmail.com>

timkicker committed:
Update Makefile

Co-authored-by: Stefan Pedratscher <pedratscher.stefan@gmail.com>

timkicker committed:
Update Makefile

Co-authored-by: Stefan Pedratscher <pedratscher.stefan@gmail.com>

timkicker commented:
Thanks a lot for the kind words and your final review!

I’ve applied the small suggestions you mentioned, should be good to merge now.
